app/src/common/shared/org/mozilla/vrbrowser/browser/engine/gecko/GeckoEngine.java

```diff
@@ -0,0 +1,255 @@
+        return mSettings;
+    }
+
+    @NotNull
+    @Override
+    public TrackingProtectionExceptionStorage getTrackingProtectionExceptionStore() {
+        return null;
+    }
+
+    @NotNull
+    @Override
+    public EngineVersion getVersion() {
+        EngineVersion version = EngineVersion.Companion.parse(MOZILLA_VERSION);
+        if (version != null) {
+            return version;
+        }
+
+        throw new IllegalStateException("Could not determine engine version");
+    }
+
+    @Override
+    public void clearData(@NotNull BrowsingData browsingData, @Nullable String host, @NotNull Function0<Unit> onSuccess, @NotNull Function1<? super Throwable, Unit> onError) {
+        long types = (long) browsingData.getTypes();
+        if (host != null) {
+            mSessionStore.getRuntime().getStorageController().clearDataFromHost(host, types).then(aVoid -> {
+                onSuccess.invoke();
+                return null;
+            }, throwable -> {
+                onError.invoke(throwable);
+                return null;
+            });
+
+        } else {
+            mSessionStore.getRuntime().getStorageController().clearData(types).then(aVoid -> {
+                onSuccess.invoke();
+                return null;
+            }, throwable -> {
+                onError.invoke(throwable);
+                return null;
+            });
+        }
+
+    }
+
+    @NotNull
+    @Override
+    public EngineSession createSession(boolean isPrivate) {
+        return new GeckoEngineSession(mContext, mSessionStore.createSession(isPrivate));
+    }
+
+    @NotNull
+    @Override
+    public EngineSessionState createSessionState(@NotNull JSONObject jsonObject) {
+        return GeckoEngineSessionState.fromJSON(jsonObject);
+    }
+
+    @NotNull
+    @Override
+    public EngineView createView(@NotNull Context context, @Nullable AttributeSet attributeSet) {
+        // We don't support Engine views creation
+        return null;
+    }
+
+    @Override
+    public void getTrackersLog(@NotNull EngineSession engineSession, @NotNull Function1<? super List<TrackerLog>, Unit> onSuccess, @NotNull Function1<? super Throwable, Unit> onError) {
+        GeckoSession session = ((GeckoEngineSession)engineSession).getGeckoSession();
+        mSessionStore.getRuntime().getContentBlockingController().getLog(session).then(logEntries -> {
+            List<ContentBlockingController.LogEntry> list = new ArrayList<>();
+            if (logEntries!= null) {
+                list = logEntries;
+            }
+            List<TrackerLog> logs = list.stream().map(this::toTrackerLog).filter(it -> !(it.getCookiesHasBeenBlocked() && it.getBlockedCategories().isEmpty() && it.getLoadedCategories().isEmpty())).collect(Collectors.toCollection(ArrayList::new));
+
+            onSuccess.invoke(logs);
+            return new GeckoResult<>();
+
+        }, throwable -> {
+            onError.invoke(throwable);
+            return new GeckoResult<>();
```

Comment:
Use GeckoResult.fromException(). Otherwise the future value will never complete.

Reply:
Fixed